python/lsst/daf/ingest/ingestCatalogTask.py

```diff
@@ -0,0 +1,576 @@
+import lsst.pex.config as pexConfig
+import lsst.pipe.base as pipeBase
+
+## \addtogroup LSST_task_documentation
+## \{
+## \page daf_ingest_IngestCatalogTask
+## \ref IngestCatalogTask_ "IngestCatalogTask"
+##      \copybrief IngestCatalogTask
+## \}
+
+class FieldFormatter(object):
+    """A class for formatting fields in an afw BaseCatalog.
+
+    This class is a container for a function that maps a Field to a MySQL type,
+    and a function that maps a field value to a literal suitable for use in a
+    MySQL INSERT/REPLACE statement.
+    """
+
+    def __init__(self, sqlTypeCallable, formatValueCallable):
+        """Store the field formatting information.
+        """
+        self.sqlTypeCallable = sqlTypeCallable
+        self.formatValueCallable = formatValueCallable
+
+    def sqlType(self, field):
+        """Return the SQL type for the given field's values.
+        """
+        return self.sqlTypeCallable(field)
+
+    def formatValue(self, value):
+        """Return a string representation of value suitable for use in an
+        INSERT/REPLACE statement (not a CSV file).  Nones are always
+        transformed to SQL NULLs.
+        """
+        if value is None:
+            return "NULL"
+        return self.formatValueCallable(value)
+
+
+def _formatNumber(formatString, number):
+    """Helper function for formatting floating point numbers. NaNs and
+    infinities are converted to NULL.
+    """
+    if math.isnan(number) or math.isinf(number):
+        return "NULL"
+    return formatString.format(number)
+
+def _formatString(string):
+    """Helper function for formatting a string. Embedded backslashes and single
+    quotes are backslash-escaped.
+    """
+    return "'" + string.replace("\\", "\\\\").replace("'", "\\'") + "'"
+
+def _formatArray(formatChar, array):
+    """Auxiliary function for formatting an array. The array elements are packed
+    into a sequence of bytes, with bytes comprising individual elements arranged
+    in little endian order. This sequence is then transformed into a MySQL
+    hexadecimal literal and returned.
+    """
+    bytestring = struct.pack("<" + str(len(array)) + formatChar, *array)
+    return "x'" + bytestring.encode('hex_codec') + "'"
+
+
+def _sqlTypeForString(field):
+    """Helper function for computing the SQL column type of a string valued field.
+    """
+    sz = field.getSize()
+    if sz > 65535:
+        # If necessary, longer strings could be ingested as TEXT.
+        raise RuntimeError("String field is too large for ingestion")
+    elif sz == 0:
+        # The field has variable size.
+        return "TEXT CHARACTER SET ascii COLLATE ascii_bin NOT NULL"
+    # A string containing trailing spaces cannot round-trip to a CHAR
+    # column and back. Therefore, we use VARCHAR even for fixed-length
+    # string fields. Also, assume strings are ASCII for now.
+    return "VARCHAR({}) CHARACTER SET ascii COLLATE ascii_bin NOT NULL".format(sz)
```

Comment:
Good catch. This comment was left over from a previous implementation. It turns out all afw table strings are variable length stuffed into fixed size storage, and that a non-zero maximum length must be specified. The use of TEXT here is in anticipation of a future afw table update that may or may not happen. I'll remove it and update the comment.